File: src/forge/github.ts
import {
    DEFAULT_PAGE,
    DEFAULT_PER_PAGE,
    type CreatePrParams,
    type Forge,
    type Issue,
    type IssueFilters,
    type Pr,
    type PrFilters,
    type WebUrlType,
} from "@/forge/forge.js";
import { exitWith } from "@/utils/debug.js";
import { AuthenticationError, ForgeHttpClient } from "@/utils/httpClient.js";

/**
 * GitHub API response for issues.
 *
 * @see https://docs.github.com/en/rest/issues/issues
 */
export interface GitHubIssueResponse {
    number: number;
    title: string;
    state: string;
    labels: Array<{ name: string }>;
    user: { login: string };
    created_at: string;
    updated_at: string;
    html_url: string;
    body?: string;
    pull_request?: unknown;
}

/**
 * GitHub API response for pull requests.
 *
 * @see https://docs.github.com/en/rest/pulls/pulls
 */
export interface GitHubPullRequestResponse {
    number: number;
    title: string;
    state: string;
    labels: Array<{ name: string }>;
    user: { login: string };
    created_at: string;
    updated_at: string;
    html_url: string;
    body: string;
    head: { ref: string };
    base: { ref: string };
    draft?: boolean;
    merged_at: string | null;
}

export class GitHubForge implements Forge {
    readonly name = "GitHub";

    private readonly _apiBase: string;
    private readonly _webBase: string;
    private readonly _repoPath: string;
    private readonly _httpClient: ForgeHttpClient;

    constructor({
        host,
        path,
        useAuth = false,
    }: {
        host: string;
        path: string;
        useAuth?: boolean;
    }) {
        this._repoPath = path;

        if (host === "github.com") {
            this._apiBase = "https://api.github.com";
            this._webBase = "https://github.com";
        } else {
            this._apiBase = `https://${host}/api/v3`;
            this._webBase = `https://${host}`;
        }

        const headers: Record<string, string> = {
            Accept: "application/vnd.github.v3+json",
        };

        if (useAuth && process.env.GITHUB_TOKEN) {
            headers["Authorization"] = `Bearer ${process.env.GITHUB_TOKEN}`;
        } else if (useAuth && !process.env.GITHUB_TOKEN) {
            exitWith(
                new AuthenticationError({
                    forgeName: "GitHub",
                    message: "Authentication enabled but GITHUB_TOKEN not set",
                }),
            );
        }

        this._httpClient = new ForgeHttpClient({
            forgeName: "GitHub",
            headers,
        });
    }

    /**
     * Lists issues from the repository.
     *
     * @param filters - Optional filter criteria
     * @returns Promise resolving to array of normalized Issue objects
     * @see https://docs.github.com/en/rest/issues/issues#list-repository-issues
     */
    async listIssues(filters?: IssueFilters): Promise<Issue[]> {
        const params = new URLSearchParams({
            state: filters?.state ?? "open",
            per_page: (filters?.perPage ?? DEFAULT_PER_PAGE).toString(),
            page: (filters?.page ?? DEFAULT_PAGE).toString(),
        });

        if (filters?.author) {
            params.set("creator", filters.author);
        }

        if (filters?.labels && filters.labels.length > 0) {
            params.set("labels", filters.labels.join(","));
        }

        const url = `${this._apiBase}/repos/${this._repoPath}/issues?${params}`;
        const issuesAndPrs =
            await this._httpClient.getJson<GitHubIssueResponse[]>(url);

        return issuesAndPrs
            .filter((item) => !item.pull_request)
            .map(mapToIssue);
    }

    /**
     * Lists pull requests from the repository.
     *
     * @param filters - Optional filter criteria
     * @returns Promise resolving to array of normalized Pr objects
     * @see https://docs.github.com/en/rest/pulls/pulls#list-pull-requests
     */
    async listPrs(filters?: PrFilters): Promise<Pr[]> {
        let apiState: string = filters?.state ?? "open";
        let filterMerged = false;

        if (filters?.state === "merged") {
            apiState = "closed";
            filterMerged = true;
        } else if (filters?.state === "all") {
            apiState = "all";
        }

        const params = new URLSearchParams({
            state: apiState,
            per_page: (filters?.perPage ?? DEFAULT_PER_PAGE).toString(),
            page: (filters?.page ?? DEFAULT_PAGE).toString(),
        });
        const url = `${this._apiBase}/repos/${this._repoPath}/pulls?${params}`;
        const pullRequests =
            await this._httpClient.getJson<GitHubPullRequestResponse[]>(url);
        let filtered = pullRequests;

        if (filterMerged) {
            filtered = filtered.filter((pr) => pr.merged_at !== null);
        } else if (filters?.state === "closed") {
            filtered = filtered.filter((pr) => pr.merged_at === null);
        }

        if (filters?.author) {
            filtered = filtered.filter(
                (pr) => pr.user.login === filters.author,
            );
        }

        if (filters?.labels && filters.labels.length > 0) {
            filtered = filtered.filter((pr) =>
                filters.labels!.every((label) =>
                    pr.labels.some((l) => l.name === label),
                ),
            );
        }

        if (filters?.draft !== undefined) {
            filtered = filtered.filter((pr) => pr.draft === filters.draft);
        }

        return filtered.map(mapToPullRequest);
    }

    /**
     * Creates a new pull request.
     *
     * @param params - Pull request creation parameters
     * @returns Promise resolving to created Pr object
     * @see https://docs.github.com/en/rest/pulls/pulls#create-a-pull-request
     */
    async createPr(params: CreatePrParams): Promise<Pr> {
        const url = `${this._apiBase}/repos/${this._repoPath}/pulls`;
        const body = {
            title: params.title,
            head: params.sourceBranch,
            base: params.targetBranch,
            body: params.body ?? "",
            draft: params.draft ?? false,
        };
        const prData =
            await this._httpClient.postJson<GitHubPullRequestResponse>(
                url,
                body,
            );

        return mapToPullRequest(prData);
    }

    /**
     * Returns the web URL for the repository or a specific page.
     *
     * @param type - URL type: "repository" (default), "issues", "prs", or "mrs"
     * @returns Full HTTPS URL to repository or specified page
     */
    getWebUrl(type?: WebUrlType): string {
        const baseUrl = `${this._webBase}/${this._repoPath}`;

        switch (type) {
            case "issues":
                return `${baseUrl}/issues`;
            case "prs":
            case "mrs":
                return `${baseUrl}/pulls`;
            default:
                return baseUrl;
        }
    }

    /**
     * Returns the git reference for fetching a pull request.
     *
     * @param prNumber - Pull request number
     * @returns Git ref spec (e.g., "pull/123/head")
     */
    getPrRef(prNumber: string): string {
        return `pull/${prNumber}/head`;
    }
}

function mapToIssue(issue: GitHubIssueResponse): Issue {
    return {
        id: issue.number.toString(),
        title: issue.title,
        state: issue.state,
        author: issue.user.login,
        url: issue.html_url,
        labels: issue.labels.map((l) => l.name),
        createdAt: new Date(issue.created_at),
        updatedAt: new Date(issue.updated_at),
    };
}

function mapToPullRequest(pr: GitHubPullRequestResponse): Pr {
    return {
        id: pr.number.toString(),
        title: pr.title,
        state: pr.merged_at ? "merged" : pr.state,
        author: pr.user.login,
        url: pr.html_url,
        labels: pr.labels.map((l) => l.name),
        createdAt: new Date(pr.created_at),
        updatedAt: new Date(pr.updated_at),
        sourceBranch: pr.head.ref,
        targetBranch: pr.base.ref,
        draft: !!pr.draft,
        mergeable: !!pr.merged_at,
    };
}

File: src/cli/forge/github.rs
use anyhow::Context;
use serde::{Deserialize, de::DeserializeOwned};

use crate::{
    cli::{
        forge::http_client::{
            self, HttpClient, IntoPaginatedResponse, PaginatedResponse, WithAuth, WithHttpStatusOk,
        },
        issue::{CreateIssueOptions, Issue, IssueState, ListIssueFilters},
        pr::{CreatePrOptions, ListPrsFilters, Pr, PrState},
    },
    git::GitRemoteData,
};

const AUTH_TOKEN: &str = "GIT_FORGE_GITHUB_TOKEN";
const AUTH_SCHEME: &str = "Bearer";

// =============================================================================
// Domain Types
// =============================================================================

/// GitHub Search API response
/// https://docs.github.com/en/rest/search/search?apiVersion=2022-11-28#search-issues-and-pull-requests
#[derive(Debug, Deserialize)]
struct GitHubSearchResponse<T> {
    items: Vec<T>,
}

impl<S, T: From<S>> IntoPaginatedResponse<T> for GitHubSearchResponse<S> {
    fn into_paginated_response(self, has_next_page: bool) -> PaginatedResponse<T> {
        PaginatedResponse::new(
            self.items.into_iter().map(Into::into).collect(),
            has_next_page,
        )
    }
}

/// GitHub API response for issues.
/// https://docs.github.com/en/rest/issues/issues
#[derive(Debug, Deserialize)]
struct GitHubIssue {
    number: u32,
    title: String,
    state: IssueState,
    labels: Vec<GitHubLabel>,
    user: GitHubUser,
    html_url: String,
}

impl From<GitHubIssue> for Issue {
    fn from(issue: GitHubIssue) -> Self {
        Issue {
            id: issue.number,
            title: issue.title,
            state: issue.state,
            author: issue.user.login,
            url: issue.html_url,
            labels: issue.labels.into_iter().map(|l| l.name).collect(),
        }
    }
}

#[derive(Debug, Deserialize)]
struct GitHubLabel {
    name: String,
}

#[derive(Debug, Deserialize)]
struct GitHubUser {
    login: String,
}

/// GitHub API response for pull requests.
/// https://docs.github.com/en/rest/pulls/pulls
#[derive(Debug, Deserialize)]
struct GitHubPullRequest {
    number: u32,
    title: String,
    state: String,
    labels: Vec<GitHubLabel>,
    user: GitHubUser,
    created_at: String,
    updated_at: String,
    html_url: String,
    draft: Option<bool>,
    merged_at: Option<String>,
}

impl From<GitHubPullRequest> for Pr {
    fn from(pr: GitHubPullRequest) -> Self {
        Pr {
            id: pr.number,
            title: pr.title,
            state: if pr.merged_at.is_some() {
                "merged".to_string()
            } else {
                pr.state
            },
            author: pr.user.login,
            url: pr.html_url,
            labels: pr.labels.into_iter().map(|l| l.name).collect(),
            created_at: pr.created_at,
            updated_at: pr.updated_at,
            draft: pr.draft.unwrap_or(false),
        }
    }
}

// =============================================================================
// Command Logic
// =============================================================================

pub fn get_issues(
    http_client: &HttpClient,
    remote: &GitRemoteData,
    api_url: Option<&str>,
    filters: &ListIssueFilters,
    use_auth: bool,
) -> anyhow::Result<PaginatedResponse<Issue>> {
    let base_url = match api_url {
        Some(url) => url,
        None => &build_api_base_url(remote),
    };
    let url = format!("{base_url}/search/issues");
    let query_string = build_issue_search_query(&remote.path, filters);

    find_items_with_search_api::<GitHubIssue, Issue>(
        http_client,
        &url,
        &query_string,
        filters.page,
        filters.per_page,
        use_auth,
    )
}

pub fn create_issue(
    http_client: &HttpClient,
    remote: &GitRemoteData,
    api_url: Option<&str>,
    options: &CreateIssueOptions,
) -> anyhow::Result<Issue> {
    let base_url = match api_url {
        Some(url) => url,
        None => &build_api_base_url(remote),
    };
    let repo_path = &remote.path;
    let url = format!("{base_url}/repos/{repo_path}/issues");
    let request_body = serde_json::json!({
        "title": options.title,
        "body": options.body,
    });

    eprintln!("Creating issue on GitHub...");

    let request = http_client
        .post(&url)
        .with_auth(true, AUTH_TOKEN, AUTH_SCHEME)?
        .header("Accept", "application/vnd.github+json")
        .json(&request_body);

    request
        .send()
        .context("Network request failed while creating issue on GitHub")?
        .with_http_status_ok()?
        .json()
        .context("Failed to parse GitHub API response")
        .map(|issue: GitHubIssue| issue.into())
}

pub fn get_prs(
    http_client: &HttpClient,
    remote: &GitRemoteData,
    api_url: Option<&str>,
    filters: &ListPrsFilters,
    use_auth: bool,
) -> anyhow::Result<PaginatedResponse<Pr>> {
    let base_url = match api_url {
        Some(url) => url,
        None => &build_api_base_url(remote),
    };
    let url = format!("{base_url}/search/issues");
    let query_string = build_pr_search_query(&remote.path, filters);

    find_items_with_search_api::<GitHubPullRequest, Pr>(
        http_client,
        &url,
        &query_string,
        filters.page,
        filters.per_page,
        use_auth,
    )
}

pub fn create_pr(
    http_client: &HttpClient,
    remote: &GitRemoteData,
    api_url: Option<&str>,
    options: &CreatePrOptions,
) -> anyhow::Result<Pr> {
    let base_url = match api_url {
        Some(url) => url,
        None => &build_api_base_url(remote),
    };
    let repo_path = &remote.path;
    let url = format!("{base_url}/repos/{repo_path}/pulls");
    let request_body = serde_json::json!({
        "title": options.title,
        "head": options.source_branch,
        "base": options.target_branch,
        "body": options.body,
        "draft": options.draft,
    });

    eprintln!("Creating pull request on GitHub...");

    let request = http_client
        .post(&url)
        .with_auth(true, AUTH_TOKEN, AUTH_SCHEME)?
        .header("Accept", "application/vnd.github+json")
        .json(&request_body);

    request
        .send()
        .context("Network request failed while creating pull request on GitHub")?
        .with_http_status_ok()?
        .json()
        .context("Failed to parse GitHub API response")
        .map(|pr: GitHubPullRequest| pr.into())
}

pub fn get_pr_ref(pr_number: u32) -> String {
    format!("pull/{pr_number}/head")
}

pub fn get_url_for_home(remote: &GitRemoteData) -> String {
    build_web_base_url(remote)
}

pub fn get_url_for_commit(remote: &GitRemoteData, commit: &str) -> String {
    format!("{}/commit/{}", build_web_base_url(remote), commit)
}

pub fn get_url_for_issue(remote: &GitRemoteData, issue_number: u32) -> String {
    format!("{}/issues/{}", build_web_base_url(remote), issue_number)
}

pub fn get_url_for_issues(remote: &GitRemoteData) -> String {
    format!("{}/issues", build_web_base_url(remote))
}

pub fn get_url_for_issue_creation(remote: &GitRemoteData) -> String {
    format!("{}/issues/new", build_web_base_url(remote))
}

pub fn get_url_for_pr(remote: &GitRemoteData, pr_number: u32) -> String {
    format!("{}/pull/{}", build_web_base_url(remote), pr_number)
}

pub fn get_url_for_prs(remote: &GitRemoteData) -> String {
    format!("{}/pulls", build_web_base_url(remote))
}

pub fn get_url_for_releases(remote: &GitRemoteData) -> String {
    format!("{}/releases", build_web_base_url(remote))
}

pub fn get_url_for_path(
    remote: &GitRemoteData,
    path: &str,
    commit: &str,
    line_number: Option<u32>,
) -> String {
    let base = build_web_base_url(remote);
    let mut url = format!("{}/blob/{}/{}", base, commit, path);

    if let Some(line) = line_number {
        url.push_str(&format!("#L{}", line));
    }

    url
}

// =============================================================================
// Private Helpers
// =============================================================================

fn build_api_base_url(remote: &GitRemoteData) -> String {
    let (host, port) = (&remote.host, remote.port);

    if host == "github.com" {
        "https://api.github.com".to_string()
    } else {
        match port {
            Some(p) => format!("https://{host}:{p}/api/v3"),
            None => format!("https://{host}/api/v3"),
        }
    }
}

fn build_web_base_url(remote: &GitRemoteData) -> String {
    let host = &remote.host;
    let path = &remote.path;

    match remote.port {
        Some(port) => format!("https://{host}:{port}/{path}"),
        None => format!("https://{host}/{path}"),
    }
}

/// https://docs.github.com/en/search-github/searching-on-github/searching-issues-and-pull-requests
fn build_issue_search_query(repo_path: &str, filters: &ListIssueFilters) -> String {
    let mut query_string = match filters.query {
        Some(query) => format!("{query} in:title in:body repo:{repo_path} is:issue"),
        None => format!("repo:{repo_path} is:issue"),
    };

    match filters.state {
        IssueState::Open => query_string.push_str(" is:open"),
        IssueState::Closed => query_string.push_str(" is:closed"),
        IssueState::All => {}
    }

    if let Some(assignee) = filters.assignee {
        query_string.push_str(" assignee:");
        query_string.push_str(assignee);
    }

    if let Some(author) = filters.author {
        query_string.push_str(" author:");
        query_string.push_str(author);
    }

    for label in filters.labels {
        query_string.push_str(" label:");
        query_string.push_str(label);
    }

    query_string
}

/// https://docs.github.com/en/search-github/searching-on-github/searching-issues-and-pull-requests
fn build_pr_search_query(repo_path: &str, filters: &ListPrsFilters) -> String {
    let mut query_string = match filters.query {
        Some(query) => format!("{query} in:title in:body repo:{repo_path} is:pr"),
        None => format!("repo:{repo_path} is:pr"),
    };

    match filters.state {
        PrState::Open => query_string.push_str(" is:open"),
        PrState::Closed => query_string.push_str(" is:closed is:unmerged"),
        PrState::Merged => query_string.push_str(" is:merged"),
        PrState::All => {}
    }

    if let Some(author) = filters.author {
        query_string.push_str(" author:");
        query_string.push_str(author);
    }

    for label in filters.labels {
        query_string.push_str(" label:");
        query_string.push_str(label);
    }

    if filters.draft {
        query_string.push_str(" draft:true");
    }

    query_string
}

fn find_items_with_search_api<T, U>(
    http_client: &HttpClient,
    url: &str,
    query_string: &str,
    page: u32,
    per_page: u32,
    use_auth: bool,
) -> anyhow::Result<PaginatedResponse<U>>
where
    T: DeserializeOwned,
    U: From<T>,
{
    let request = http_client
        .get(url)
        .with_auth(use_auth, AUTH_TOKEN, AUTH_SCHEME)?
        .header("Accept", "application/vnd.github+json")
        .query(&[("q", query_string)])
        .query(&[("page", page)])
        .query(&[("per_page", per_page)]);

    let response = request
        .send()
        .context("Failed to fetch items from GitHub Search API")?
        .with_http_status_ok()?;

    let has_next_page = http_client::has_next_link_header(&response);

    response
        .json()
        .context("Failed to parse GitHub Search API response")
        .map(|res: GitHubSearchResponse<T>| res.into_paginated_response(has_next_page))
}
